std/assembly/util/hash.ts

```diff
@@ -10,63 +8,137 @@ export function HASH<T>(key: T): u32 {
     if (sizeof<T>() == 8) return hash64(u64(key));
   }
   return unreachable();
 }
 
-// FNV-1a 32-bit as a starting point, see: http://isthe.com/chongo/tech/comp/fnv/
+// XXHash 32-bit as a starting point, see: https://cyan4973.github.io/xxHash
 
+// primes
 // @ts-ignore: decorator
-@inline const FNV_OFFSET: u32 = 2166136261;
-
+@inline const XXH64_P1: u64 = 11400714785074694791;
+// @ts-ignore: decorator
+@inline const XXH64_P2: u64 = 14029467366897019727;
+// @ts-ignore: decorator
+@inline const XXH64_P3: u64 = 1609587929392839161;
 // @ts-ignore: decorator
-@inline const FNV_PRIME: u32 = 16777619;
+@inline const XXH64_P4: u64 = 9650029242287828579;
+// @ts-ignore: decorator
+@inline const XXH64_P5: u64 = 2870177450012600261;
+// @ts-ignore: decorator
+@inline const XXH64_SEED: u64 = 0;
 
-function hash8(key: u32): u32 {
-  return (FNV_OFFSET ^ key) * FNV_PRIME;
+// @ts-ignore: decorator
+@inline
+function hash32(key: u32, len: u64 = 4): u64 {
+  var h: u64 = XXH64_SEED + XXH64_P5 + len;
+  h ^= u64(key) * XXH64_P1;
+  h  = rotl(h, 23) * XXH64_P2 + XXH64_P3;
+  h ^= h >> 33;
+  h *= XXH64_P2;
+  h ^= h >> 29;
+  h *= XXH64_P3;
+  h ^= h >> 32;
+  return h;
 }
 
-function hash16(key: u32): u32 {
-  var v = FNV_OFFSET;
-  v = (v ^ ( key        & 0xff)) * FNV_PRIME;
-  v = (v ^ ( key >>  8        )) * FNV_PRIME;
-  return v;
+// @ts-ignore: decorator
+@inline
+function hash64(key: u64): u64 {
+  var h: u64 = XXH64_SEED + XXH64_P5 + 8;
+  h ^= rotl(key * XXH64_P2, 31) * XXH64_P1;
+  h  = rotl(h, 27) * XXH64_P1 + XXH64_P4;
+  h ^= h >> 33;
+  h *= XXH64_P2;
+  h ^= h >> 29;
+  h *= XXH64_P3;
+  h ^= h >> 32;
+  return h;
 }
 
-function hash32(key: u32): u32 {
-  var v = FNV_OFFSET;
-  v = (v ^ ( key        & 0xff)) * FNV_PRIME;
-  v = (v ^ ((key >>  8) & 0xff)) * FNV_PRIME;
-  v = (v ^ ((key >> 16) & 0xff)) * FNV_PRIME;
-  v = (v ^ ( key >> 24        )) * FNV_PRIME;
-  return v;
+// @ts-ignore: decorator
+@inline
+function mix1(h: u64, key: u64): u64 {
+  return rotl(h + key * XXH64_P2, 31) * XXH64_P1;
 }
 
-function hash64(key: u64): u32 {
-  var l = <u32> key;
-  var h = <u32>(key >>> 32);
-  var v = FNV_OFFSET;
-  v = (v ^ ( l        & 0xff)) * FNV_PRIME;
-  v = (v ^ ((l >>  8) & 0xff)) * FNV_PRIME;
-  v = (v ^ ((l >> 16) & 0xff)) * FNV_PRIME;
-  v = (v ^ ( l >> 24        )) * FNV_PRIME;
-  v = (v ^ ( h        & 0xff)) * FNV_PRIME;
-  v = (v ^ ((h >>  8) & 0xff)) * FNV_PRIME;
-  v = (v ^ ((h >> 16) & 0xff)) * FNV_PRIME;
-  v = (v ^ ( h >> 24        )) * FNV_PRIME;
-  return v;
+// @ts-ignore: decorator
+@inline
+function mix2(h: u64, s: u64): u64 {
+  return (h ^ (rotl(s, 31) * XXH64_P1)) * XXH64_P1 + XXH64_P4;
 }
 
-function hashStr(key: string): u32 {
-  var v = FNV_OFFSET;
-  if (key !== null) {
-    for (let i: usize = 0, k: usize = key.length << 1; i < k; ++i) {
-      v = (v ^ <u32>load<u8>(changetype<usize>(key) + i)) * FNV_PRIME;
+// @ts-ignore: decorator
+@inline
+function hashStr(key: string): u64 {
```

Comment:
Do you have a reference for the code in this function?

Reply:
It's loosely based on this: https://github.com/Cyan4973/xxHash/blob/dev/xxhash.h